src/cdk/a11y/input-modality/input-modality-detector.ts

```diff
@@ -0,0 +1,187 @@
+ * with a screen reader is akin to visually scanning a page, and should not be interpreted as actual
+ * user input interaction.
+ *
+ * When a user is not navigating but *interacting* with a screen reader, this service's behavior is
+ * largely undefined and depends on the events fired. For example, in VoiceOver, no keyboard events
+ * are fired when performing an element's default action via Caps Lock + Space, thus no input
+ * modality is detected.
+ */
+@Injectable({ providedIn: 'root' })
+export class InputModalityDetector implements OnDestroy {
+  /** Emits when the input modality changes. */
+  readonly inputModalityChange: Observable<InputModality>;
+
+  /** Returns the last detected / current input modality. */
+  get inputModality(): InputModality {
+    return this._inputModality;
+  }
+
+  /** The underlying ReplaySubject that emits input modality changes. */
+  private readonly _change = new ReplaySubject<InputModality>(1);
+
+  /** The current / last detected input modality by this service. */
+  private _inputModality: InputModality = null;
+
+  /** Options for this InputModalityDetector. */
+  private readonly _options: InputModalityDetectorOptions;
+
+  /**
+   * The timestamp of the last touch input modality. Used to determine whether mousedown events
+   * should be attributed to mouse or touch.
+   */
+  private _lastTouchMs = 0;
+
+  /**
+   * Handles keyboard events. Must be an arrow function in order to preserve the context when it
+   * gets bound.
+   */
+  private _onKeydown = (event: KeyboardEvent) => {
+    // If this is one of the keys we should ignore, then ignore it and don't update the input
+    // modality to keyboard.
+    if (this._options?.ignoreKeys?.some(key => key === event.key)) { return; }
+
+    this._updateInputModality('keyboard');
+  }
+
+  /**
+   * Handles mouse events. Must be an arrow function in order to preserve the context when it gets
+   * bound.
+   */
+  private _onMousedown = (event: MouseEvent) => {
+    if (isFakeMousedownFromScreenReader(event)) { return; }
+
+    // Touches trigger both touch and mouse events, so we need to distinguish between mouse events
+    // that were triggered via mouse vs touch. To do so, check if the mouse event occurs closely
+    // after the previous touch event.
+    if (Date.now() - this._lastTouchMs < TIME_AFTER_TOUCH_MS) { return; }
+
+    this._updateInputModality('mouse');
+  }
+
+  /**
+   * Handles touchstart events. Must be an arrow function in order to preserve the context when it
+   * gets bound.
+   */
+  private _onTouchstart = (event: TouchEvent) => {
+    if (isFakeTouchstartFromScreenReader(event)) { return; }
+
+    // Store the timestamp of this touch event, as it's used to distinguish between mouse events
+    // triggered via mouse vs touch.
+    this._lastTouchMs = Date.now();
+
+    this._updateInputModality('touch');
+  }
+
+  constructor(
+    platform: Platform,
+    @Inject(DOCUMENT) document: Document,
```

Comment:
AFAIK there are some issues in ViewEngine with typing constructor parameters as `Document`. The way we work around it in other places is to type the parameter as `any` and then cast it to `Document` afterwards.

Reply:
I did notice this when looking at other services. Do you recall how to reproduce these issues? I'd prefer to be able to reproduce the issue to confirm the `any` + cast is actually needed, but if you're confident, then I'll just go ahead and make the change.

Reply:
I think that we were running into it in our Angular Universal integration tests. There's a readme under `src\universal-app\DEBUG.md`.

Reply:
I added the `InputModalityDetector` to the universal kitchen sink app as-is, and ran `bazel test //src/universal-app:server_test --test_output=all`. The target passed, and I didn't see any relevant console warnings/logs. Is there something in particular I should be looking for?